# Data Cleaning
### Source: MS FLow as of 20Sept2024.xlsx

In [180]:
import pandas as pd
import numpy as np

df = pd.read_excel('/content/Data from MS FLow as of 20Sept2024.xlsx')
df.head(30)

Title       Contract Type  \
0   Letter of Agreement to extend grant funding of...           Agreement   
1   UN -UN Agreement UNCDF with FAO- Tanzania Cook...           Agreement   
2   Termination of the Grant Agreement No. 15559 b...  Contract Amendment   
3   Letter of Agreement with a Government Partner:...           Agreement   
4   Statement of Intent between the UNCDF and Comm...                 MOU   
5                     MOU between UNCDF and the UNOPS                 MOU   
6                              PBPA Bangladesh LoGIC                  PBA   
7   Low Value Performance-Based Payment Agreement ...                 PBA   
8   No-cost extension of the Global Local Finance ...                 PBA   
9   Signature and approval of new Low Value PBPA L...                 PBA   
10  Request for clearance of EU agreement clearanc...           Agreement   
11  Signature and approval of the LOA MINISTRY OF ...           Agreement   
12  Project document for the Municipal Investment ...                 PBA   
13  UNDP Brussles Clearance Slip for new contribut...           Agreement   
14  Netherlands LMFTF Amendment#2 and Disbursement...  Contract Amendment   
15  FAO-UNCDF amendment (review of the disbursemen...  Contract Amendment   
16  3-month no cost extension (till 31 December 20...  Contract Amendment   
17  No Cost Extension of the Grant Agreement with ...           Agreement   
18  MOU between UNCDF and the Ministry of Environm...                 MOU   
19   E-KYC program - signature of Agreement Amendment  Contract Amendment   
20  Signature and approval of the LOA Direction Gé...           Agreement   
21  Financing Agreement between the United Nations...  Contract Amendment   
22  new LOA Gaza-MEF, Mozambique, for signature of...           Agreement   
23  UNCDF WDFI Coalition Ethiopia- No Cost Extensi...           Agreement   
24  Request for Amendment to Contract Starting Dat...  Contract Amendment   
25  Endoresement: French contribution to Policy Ac...           Agreement   
26  Transfer of Ownership Agreement between GIZ an...           Agreement   
27                   Burundi UNJP -UNCDF/WFP/FAO/UNDP           Agreement   
28  Request for ES signature - Signing of Amendmen...  Contract Amendment   
29  Joint UN MOU as a non-exclusive cooperation, f...                 MOU   

               Created          Created By Practice Approval  \
0  2024-09-19 07:58:07      Aline Kayitesi           Approve   
1  2024-09-19 00:18:09    Abraham Byamungu           Approve   
2  2024-09-18 16:50:47      Aline Kayitesi           Approve   
3  2024-09-18 11:50:22      Aline Kayitesi           Approve   
4  2024-09-16 11:52:45     Shriya Sundaram          Approve    
5  2024-09-16 11:44:47     Shriya Sundaram          Approve    
6  2024-09-13 02:42:42    Kulrisa Shayavee           Approve   
7  2024-09-12 08:24:36    Kulrisa Shayavee           Approve   
8  2024-09-11 09:17:23      Aline Kayitesi           Approve   
9  2024-09-10 22:48:06    Kulrisa Shayavee           Approve   
10 2024-09-09 06:57:38    Kulrisa Shayavee           Approve   
11 2024-09-09 06:08:25    Kulrisa Shayavee           Approve   
12 2024-09-06 08:41:35      Aline Kayitesi           Approve   
13 2024-09-05 05:45:56    Kulrisa Shayavee               NaN   
14 2024-09-04 12:39:56      Jessieca Aduca           Approve   
15 2024-09-03 03:31:00  Francois Coupienne          Approve    
16 2024-08-30 04:04:47     Francesca Cioni          Approve    
17 2024-08-27 09:19:59      Aline Kayitesi           Approve   
18 2024-08-27 03:24:43       Rahayu Maitri               NaN   
19 2024-08-26 06:56:00   Melania Giubbilei          Approve    
20 2024-08-22 23:02:33    Kulrisa Shayavee           Approve   
21 2024-08-22 08:19:42        Meron Tamrat          Approve    
22 2024-08-05 21:02:00    Kulrisa Shayavee           Approve   
23 2024-07-22 08:07:05     Belinda Uriarte          Approve    
24 2024-07-15 09:18:34    Oranat Prasanpot           Approve

In [181]:
#Check number of rows and columns
df.shape

(2695, 14)

In [182]:
#check datatype for all columns

df.dtypes


Title                                       object
Contract Type                               object
Created                             datetime64[ns]
Created By                                  object
Practice Approval                           object
Partnership Approval                        object
Head of PPC Approval                        object
Partnership Reviewer                        object
Practice Reviewer                           object
Attached signed agreement                  float64
ES/DES Review                               object
Sign-off status (entered by PPC)            object
Item Type                                   object
Path                                        object
dtype: object

## Dealing With Missing Values

In [183]:
#Drop missing values from "Head of PPC Approval"
df = df.dropna(subset=['Head of PPC Approval'])
df.head()

Title       Contract Type  \
2   Termination of the Grant Agreement No. 15559 b...  Contract Amendment   
4   Statement of Intent between the UNCDF and Comm...                 MOU   
5                     MOU between UNCDF and the UNOPS                 MOU   
10  Request for clearance of EU agreement clearanc...           Agreement   
15  FAO-UNCDF amendment (review of the disbursemen...  Contract Amendment   

               Created          Created By Practice Approval  \
2  2024-09-18 16:50:47      Aline Kayitesi           Approve   
4  2024-09-16 11:52:45     Shriya Sundaram          Approve    
5  2024-09-16 11:44:47     Shriya Sundaram          Approve    
10 2024-09-09 06:57:38    Kulrisa Shayavee           Approve   
15 2024-09-03 03:31:00  Francois Coupienne          Approve    

   Partnership Approval Head of PPC Approval Partnership Reviewer  \
2               Approve              Approve       Jessieca Aduca   
4               Approve              Approve       Jessieca Aduca   
5               Approve              Approve       Jessieca Aduca   
10              Approve              Approve       Jessieca Aduca   
15              Approve              Approve       Jessieca Aduca   

           Practice Reviewer  Attached signed agreement ES/DES Review  \
2   Jenifer Bukokhe Wakhungu                        0.0           NaN   
4              Maria Perdomo                        0.0           NaN   
5              Maria Perdomo                        0.0           NaN   
10  Jenifer Bukokhe Wakhungu                        0.0           NaN   
15             Maria Perdomo                        1.0           NaN   

   Sign-off status (entered by PPC) Item Type  \
2                               NaN      Item   
4                               NaN      Item   
5                               NaN      Item   
10                              NaN      Item   
15                              NaN      Item   

                                                 Path  
2   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...  
4   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...  
5   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...  
10  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...  
15  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...

In [184]:
df.shape

(762, 14)

762 rows left after removing missing values from "`Head of PPC Approval`" column

## Dealing with duplicates

In [185]:
# Duplicate the 'Title' and 'Created' columns
df['Title_Duplicate'] = df['Title']
df['Created_Duplicate'] = df['Created']

# Merge the duplicated columns into a new column (e.g., 'Merged_Title_Created')
df['Merged_Title_Created'] = df['Title'].astype(str) + ' - ' + df['Created'].astype(str)

# Display the updated DataFrame (Optional)
df.head()


<ipython-input-185-3ef44a3b1a83>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Title_Duplicate'] = df['Title']
<ipython-input-185-3ef44a3b1a83>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Created_Duplicate'] = df['Created']
<ipython-input-185-3ef44a3b1a83>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user

Title       Contract Type  \
2   Termination of the Grant Agreement No. 15559 b...  Contract Amendment   
4   Statement of Intent between the UNCDF and Comm...                 MOU   
5                     MOU between UNCDF and the UNOPS                 MOU   
10  Request for clearance of EU agreement clearanc...           Agreement   
15  FAO-UNCDF amendment (review of the disbursemen...  Contract Amendment   

               Created          Created By Practice Approval  \
2  2024-09-18 16:50:47      Aline Kayitesi           Approve   
4  2024-09-16 11:52:45     Shriya Sundaram          Approve    
5  2024-09-16 11:44:47     Shriya Sundaram          Approve    
10 2024-09-09 06:57:38    Kulrisa Shayavee           Approve   
15 2024-09-03 03:31:00  Francois Coupienne          Approve    

   Partnership Approval Head of PPC Approval Partnership Reviewer  \
2               Approve              Approve       Jessieca Aduca   
4               Approve              Approve       Jessieca Aduca   
5               Approve              Approve       Jessieca Aduca   
10              Approve              Approve       Jessieca Aduca   
15              Approve              Approve       Jessieca Aduca   

           Practice Reviewer  Attached signed agreement ES/DES Review  \
2   Jenifer Bukokhe Wakhungu                        0.0           NaN   
4              Maria Perdomo                        0.0           NaN   
5              Maria Perdomo                        0.0           NaN   
10  Jenifer Bukokhe Wakhungu                        0.0           NaN   
15             Maria Perdomo                        1.0           NaN   

   Sign-off status (entered by PPC) Item Type  \
2                               NaN      Item   
4                               NaN      Item   
5                               NaN      Item   
10                              NaN      Item   
15                              NaN      Item   

                                                 Path  \
2   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
4   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
5   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
10  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
15  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   

                                      Title_Duplicate   Created_Duplicate  \
2   Termination of the Grant Agreement No. 15559 b... 2024-09-18 16:50:47   
4   Statement of Intent between the UNCDF and Comm... 2024-09-16 11:52:45   
5                     MOU between UNCDF and the UNOPS 2024-09-16 11:44:47   
10  Request for clearance of EU agreement clearanc... 2024-09-09 06:57:38   
15  FAO-UNCDF amendment (review of the disbursemen... 2024-09-03 03:31:00   

                                 Merged_Title_Created  
2   Termination of the Grant Agreement No. 15559 b...  
4   Statement of Intent between the UNCDF and Comm...  
5   MOU between UNCDF and the UNOPS - 2024-09-16 1...  
10  Request for clearance of EU agreement clearanc...  
15  FAO-UNCDF amendment (review of the disbursemen...

In [186]:
# Remove duplicates
df = df.drop_duplicates(subset=['Merged_Title_Created'])
df.shape

(762, 17)

In [187]:
# Recheck for duplicates
df['Title_duplicate'] = df.duplicated(subset=['Title'], keep=False)
df_true = df[df['Title_duplicate'] == True]
df_true.shape

(8, 18)

In [188]:
df_true.head(9)

Title       Contract Type  \
150   Request for agreement clearance - Signature of...           Agreement   
192   Request for agreement clearance - Signature of...           Agreement   
808   For clearance and signature, new contribution ...           Agreement   
815   For clearance and signature, new contribution ...           Agreement   
1700        Exchange of Letters between UNCDF and UNFPA           Agreement   
1789        Exchange of Letters between UNCDF and UNFPA  Contract Amendment   
2236        MOU between UNCDF and Women's World Banking                 MOU   
2254        MOU between UNCDF and Women's World Banking                 MOU   

                 Created             Created By Practice Approval  \
150  2023-12-27 06:04:58       Kulrisa Shayavee           Approve   
192  2023-11-24 17:26:40       Kulrisa Shayavee           Approve   
808  2022-11-28 20:23:32       Kulrisa Shayavee           Approve   
815  2022-11-25 17:12:56       Kulrisa Shayavee           Approve   
1700 2021-06-22 11:45:10        Janine Ouattara          Approve    
1789 2021-05-04 07:06:09        Janine Ouattara          Approve    
2236 2020-07-03 04:25:11  Nandini Harihareswara          Approve    
2254 2020-06-29 19:05:22  Nandini Harihareswara          Approve    

     Partnership Approval Head of PPC Approval Partnership Reviewer  \
150               Approve              Approve       Jessieca Aduca   
192               Approve              Approve       Jessieca Aduca   
808               Approve              Approve       Jessieca Aduca   
815               Approve              Approve     Abhisheik Dhawan   
1700              Approve              Approve       Jessieca Aduca   
1789              Approve              Approve       Jessieca Aduca   
2236              Approve              Approve     Abhisheik Dhawan   
2254              Approve               Reject     Abhisheik Dhawan   

             Practice Reviewer  Attached signed agreement ES/DES Review  \
150   Jenifer Bukokhe Wakhungu                        1.0           NaN   
192   Jenifer Bukokhe Wakhungu                        0.0           NaN   
808   Jenifer Bukokhe Wakhungu                        0.0           NaN   
815   Jenifer Bukokhe Wakhungu                        0.0           NaN   
1700               John Tucker                        1.0           NaN   
1789               John Tucker                        0.0           NaN   
2236               John Tucker                        NaN           NaN   
2254               John Tucker                        NaN           NaN   

     Sign-off status (entered by PPC) Item Type  \
150                               NaN      Item   
192                               NaN      Item   
808                               NaN      Item   
815                               NaN      Item   
1700                              NaN      Item   
1789                              NaN      Item   
2236                              NaN      Item   
2254                              NaN      Item   

                                                   Path  \
150   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
192   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
808   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
815   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
1700  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
1789  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
2236  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   
2254  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...   

                                        Title_Duplicate   Created_Duplicate  \
150   Request for agreement clearance - Signature of... 2023-12-27 06:04:58   
192   Request for agreement clearance - Signature of... 2023-11-24 17:26:40   
808   For clearance and signature, new contribution ... 2022-11-28 20:23:32   
815   For clearance and signature, new contribution ... 2022-11-25 17:12:56   
1700        E

8 Duplicate titles with different "`Created`" date identify
(Clarify with Fernando and Jessieca to delete or leave it be)

In [189]:
# Drop Merged_Title_Created, Created_Duplicate
df = df.drop(['Merged_Title_Created','Created_Duplicate','Title_Duplicate'], axis=1)
df_true = df_true.drop(['Merged_Title_Created','Created_Duplicate','Title_Duplicate'], axis=1)

In [190]:
# Attached signed agreement ='0.0'

df_no_attached = df[df['Attached signed agreement'] == 0.0]
df_no_attached.head()


Title       Contract Type  \
2   Termination of the Grant Agreement No. 15559 b...  Contract Amendment   
4   Statement of Intent between the UNCDF and Comm...                 MOU   
5                     MOU between UNCDF and the UNOPS                 MOU   
10  Request for clearance of EU agreement clearanc...           Agreement   
16  3-month no cost extension (till 31 December 20...  Contract Amendment   

               Created        Created By Practice Approval  \
2  2024-09-18 16:50:47    Aline Kayitesi           Approve   
4  2024-09-16 11:52:45   Shriya Sundaram          Approve    
5  2024-09-16 11:44:47   Shriya Sundaram          Approve    
10 2024-09-09 06:57:38  Kulrisa Shayavee           Approve   
16 2024-08-30 04:04:47   Francesca Cioni          Approve    

   Partnership Approval Head of PPC Approval Partnership Reviewer  \
2               Approve              Approve       Jessieca Aduca   
4               Approve              Approve       Jessieca Aduca   
5               Approve              Approve       Jessieca Aduca   
10              Approve              Approve       Jessieca Aduca   
16              Approve              Approve       Jessieca Aduca   

           Practice Reviewer  Attached signed agreement ES/DES Review  \
2   Jenifer Bukokhe Wakhungu                        0.0           NaN   
4              Maria Perdomo                        0.0           NaN   
5              Maria Perdomo                        0.0           NaN   
10  Jenifer Bukokhe Wakhungu                        0.0           NaN   
16              Henri Dommel                        0.0           NaN   

   Sign-off status (entered by PPC) Item Type  \
2                               NaN      Item   
4                               NaN      Item   
5                               NaN      Item   
10                              NaN      Item   
16                              NaN      Item   

                                                 Path  Title_duplicate  
2   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...            False  
4   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...            False  
5   sites/UNCDFAgreementMOUClearance/Lists/Agreeme...            False  
10  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...            False  
16  sites/UNCDFAgreementMOUClearance/Lists/Agreeme...            False

In [191]:
df_no_attached.shape

(261, 15)

261 data have missing attachments

In [192]:
#convert to excel worksheet

with pd.ExcelWriter('output.xlsx') as writer:
  df.to_excel(writer, sheet_name='Main', index=False)
  df_true.to_excel(writer, sheet_name='Duplicates', index=False)
  df_no_attached.to_excel(writer, sheet_name='Missing Attachments', index=False)

from google.colab import files
files.download('output.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>